protocol/messenger_handler.go

```diff
@@ -1027,58 +1027,75 @@ importMessageArchivesLoop:
 			// wait for all archives to be processed first
 			downloadedArchiveID := archiveIDsToImport[0]
 
-			messagesToHandle, err := m.communitiesManager.ExtractMessagesFromHistoryArchives(id, []string{downloadedArchiveID})
-			if err != nil {
-				m.communitiesManager.LogStdout("failed to extract history archive messages", zap.Error(err))
-				continue
-			}
-
-			importedMessages := make(map[transport.Filter][]*types.Message, 0)
-			otherMessages := make(map[transport.Filter][]*types.Message, 0)
+			m.communitiesManager.ExtractMessagesFromHistoryArchives(id, []string{downloadedArchiveID}, func(archiveMessages []*protobuf.WakuMessage, err error) {
```

Comment:
This is a very large inline function, and as far as I can see the only parameter it consumes from outside the function scope is `downloadedArchiveID`, which could be passed as a parameter.

Perhaps this logic should be made an explicit function of `Messenger`, you can still pass it in the same way.

`m.communitiesManager.ExtractMessagesFromHistoryArchives(id, []string{downloadedArchiveID}, m.yourPostProcessingFunction`)

Or alternatively return `archiveMessages []*protobuf.WakuMessage, e error` from `ExtractMessagesFromHistoryArchives ` and then pass those vars to the new function. I can see that the cause of this requirement in your code is because you require structs and functionality from outside of the scope of `communitiesManager`.

Returning the result of the new function will keep a clean separation of concerns and will be more maintainable. What do you think?

Reply:
> Or alternatively return archiveMessages []*protobuf.WakuMessage, e error from ExtractMessagesFromHistoryArchives  and then pass those vars to the new function. I can see that the cause of this requirement in your code is because you require structs and functionality from outside of the scope of communitiesManager.

This was actually the case before, but since we're not processing message archives in chunks, instead of loading all archive messages into memory that we're interested in, we have to provide a callback (or use some channel gymnastics) to let the function process a subset of messages before it loads the next chunk.

> Perhaps this logic should be made an explicit function of Messenger, you can still pass it in the same way.

Very happy to make this change. In fact I started out like this and decided to keep it inline because there was no other place where this functionality is used.

Will update this!

Reply:
> This was actually the case before, but since we're now processing message archives in chunks, instead of loading all archive messages into memory that we're interested in, we have to provide a callback (or use some channel gymnastics) to let the function process a subset of messages before it loads the next chunk.

Actually, now that we change this to only expect a single archive ID anyways, we don't need the `done` callback anymore, so what you're saying makes sense, we can now go back to returning `messages, err`

Thanks for pointing that out!